### PR TITLE
SSSP threading

### PR DESCRIPTION
**Pull request by orrrrtem**
# Added threading for SSSP.
Features and details:
- Reimplemented algorithm flow to avoid code duplication.
- Non-lock-free atomic operations are implemented via oneapi::detail::mutex and oneapi::detail::scoped lock.
- Lock-free atomic operations on double and distances+predcessors (int32) are implemented via reinterpret cast to int64.
- Other lock-free atomic operations are based on std atomic implementation from  STL
- Fixed inner initializing while constructing and destructing trivial class in vector container

### CODE REVIEW

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 1]
```diff
@@ -694,11 +694,6 @@ class shortest_paths_test {
         if (true_distances.size() != distances.size()) {
             return false;
         }
-        for (size_t index = 0; index < true_distances.size(); ++index) {
-            if (!compare_distances(true_distances[index], distances[index])) {
-                return false;
-            }
-        }
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:701]: to revert

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/service.hpp [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    data_to_relax_base(std::int64_t vertex_count,
+                       Vertex source,
+                       relaxing_data_type max_dist,
+                       byte_alloc_iface* alloc_ptr)
+            : vertex_count(vertex_count),
+              atomic_value_allocator(alloc_ptr) {
+        relaxing_data = allocate(atomic_value_allocator, vertex_count);
+        relaxing_data = new (relaxing_data) atomic_type[vertex_count]();
+        dal::detail::threader_for(vertex_count, vertex_count, [&](std::int64_t i) {
+            store_by_index(i, max_dist);
```
Comment (orazve) [inline cpp/oneapi/dal/algo/shortest_paths/backend/cpu/service.hpp:84]: won't you construct the value in parallel? and avoid code in 82 row
`new (relaxing_data[i]) atomic_type(max_dist);`

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp [hunk 3]
```diff
@@ -17,12 +17,16 @@
 
 #pragma once
 
+#include <atomic>
```
Comment (orazve) [inline cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp:20]: it is already included in service.hpp

**Review by orazve — APPROVED** [review]

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp [hunk 4]
```diff
@@ -65,6 +68,28 @@ inline void relax_edges(const Topology& t,
     }
 }
 
+template <typename Topology, typename EdgeValue, typename BinsVector>
+inline void relax_edges(const Topology& t,
```
[the remark this answers is not in the document]
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp]: we can keep relax_edges without suffix as it is the only available configuration (no sequential version)

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp [hunk 5]
```diff
@@ -17,19 +17,169 @@
...
 using namespace oneapi::dal::preview::backend;
 
+namespace mode {
+struct distances;
+struct distances_predecessors;
+} // namespace mode
+
+template <typename T1, typename T2, typename T1_, typename T2_>
+inline constexpr bool is_class_params_the_same_v =
+    std::conjunction_v<std::is_same<T1, T2>, std::is_same<T1, T1_>, std::is_same<T2, T2_>>;
```
[the remark this answers is not in the document]
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp]: No misses due to the transitivity property of is_same.

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp [hunk 6]
```diff
@@ -17,19 +17,169 @@
...
+        typename V = value_type,
+        typename AV = atomic_value_type,
+        std::enable_if_t<is_not_class_params_the_same_v<V, AV, value_type, atomic_value_type> &&
+                             std::is_same_v<value_type, double> &&
+                             std::is_same_v<atomic_value_type, std::int64_t>,
+                         bool> = true>
+    inline bool compare_exchange_strong(Vertex u, double& old_value, double new_value) {
+        double* old_value_ptr = &old_value;
+        double* new_value_ptr = &new_value;
+        std::int64_t old_value_int_representation = *reinterpret_cast<std::int64_t*>(old_value_ptr);
```
[the remark this answers is not in the document]
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp]: ci fails with gcc werror deref...

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/service.hpp (hunk 2 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/service.hpp:84]: Thanks. Need to update as such initialization atomic<T> obj(t) is not atomic and we will reduce the amount of atomics ops

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp (hunk 3 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp:20]: /intelci: run

**Pull request merged**